code/datums/components/anti_magic.dm

```diff
@@ -1,54 +1,125 @@
+	RegisterSignal(equipper, COMSIG_MOB_RECEIVE_MAGIC, .proc/block_receiving_magic)
+	RegisterSignal(equipper, COMSIG_MOB_RESTRICT_MAGIC, .proc/restrict_casting_magic)
+	equipper.update_action_buttons()
+
+	// Check to see if we have any spells that are blocked due to antimagic
+	for(var/datum/action/spell_action/spell/magic_action in equipper.actions)
+		var/obj/effect/proc_holder/spell/magic_spell = magic_action.target
+		if(antimagic_flags & magic_spell.antimagic_flags)
+			to_chat(equipper, span_warning("[parent] is interfering with your ability to cast magic!"))
+			break
+
 
 /datum/component/anti_magic/proc/on_drop(datum/source, mob/user)
 	SIGNAL_HANDLER
 
 	UnregisterSignal(user, COMSIG_MOB_RECEIVE_MAGIC)
+	UnregisterSignal(user, COMSIG_MOB_RESTRICT_MAGIC)
+	user.update_action_buttons()
 
-/datum/component/anti_magic/proc/protect(datum/source, mob/user, _magic, _holy, _psychic, chargecost, self, list/protection_sources)
+/datum/component/anti_magic/proc/block_receiving_magic(datum/source, mob/user, casted_magic_flags, charge_cost)
 	SIGNAL_HANDLER
 
-	if(((_magic && magic) || (_holy && holy) || (_psychic && psychic)) && (!self || blocks_self))
-		protection_sources += parent
-		reaction?.Invoke(user, chargecost, parent)
-		charges -= chargecost
-		if(charges <= 0)
-			expire?.Invoke(user, parent)
-			qdel(src)
-		return COMPONENT_BLOCK_MAGIC
+	// disclaimer - All anti_magic sources will be drained a charge_cost
+	if(casted_magic_flags & antimagic_flags)
+		// im a programmer not shakesphere to the future grammar nazis that come after me for this
+		var/visible_subject = ismob(parent) ? "[user.p_they()]" : "[parent]"
+		var/self_subject = ismob(parent) ? "you" : "[parent]"
+		if(casted_magic_flags & antimagic_flags & MAGIC_RESISTANCE)
+			user.visible_message(span_warning("[user] pulses red as [visible_subject] absorbs magic energy!"), \
+			span_userdanger("An intense magical aura pulses around [self_subject] as it dissipates into the air!"))
+		else if(casted_magic_flags & antimagic_flags & MAGIC_RESISTANCE_HOLY)
+			user.visible_message(span_warning("[user] starts to glow as [visible_subject] emits a halo of light!"), \
+			span_userdanger("A feeling of warmth washes over [self_subject] as rays of light surround your body and protect you!"))
+		else if(casted_magic_flags & antimagic_flags & MAGIC_RESISTANCE_MIND)
+			user.visible_message(span_warning("[user] forehead shines as [visible_subject] repulses magic from their mind!"), \
+			span_userdanger("A feeling of cold splashes on [self_subject] as your forehead reflects magic targeting your mind!"))
+
+		if(ismob(parent))
+			return TRUE
+
+		var/has_limited_charges = !(charges == INFINITY)
+		var/charge_was_drained = charge_cost > 0
+		if(has_limited_charges && charge_was_drained)
+			drain_antimagic?.Invoke(user, parent)
+			charges -= charge_cost
+			if(charges <= 0)
+				expiration?.Invoke(user, parent)
+				qdel(src)
+		return TRUE
+	return FALSE
```

Comment:
Signal procs generally should return bitflags and not booleans 

So, something like `COMPONENT_MAGIC_BLOCKED`

Reply:
Applied with 76995de9625789fe9b43ce2f18e9d2328eaaaa4f